### DISCUSSION TITLE
saving YALMIP objects in MATLAB

### DISCUSSION THREAD

**Discussion by gayanlanke** [yalmip/YALMIP#1090] [Optimization modelling] [answered]
I am trying to obtain the convex hull of a set of convex polygons as follows.

```
clc;clear

sdpvar x y

F1 = [-1 <= x <= 1, -1 <= y <= 1];  
F2 = [-3 <= x <= -2, -2 <= y <= 2];
F3 = [-5 <= x <= -4, -3 <= y <= 3];
F4 = [-9 <= x <= -6, -5 <= y <= 5];

H=hull(F1,F2,F3,F4);  %% convex hull
```

Actually, I am performing the `hull` operation in an HPC for computational reasons. Thereafter, I need to export `H` to my local computer for plotting purposes (the HPC does not support graphical representation in MATLAB).

Although I tried saving the workspace in MATLAB with `save('myfile.mat')`, when I export the .mat file, `H` seems to be empty.

Moreover, I am getting the following warning message when trying the save the workspace:

```
 In constraint/saveobj (line 3)
  In time_series_envelopes_plots_v3 (line 91)
Warning: YALMIP objects cannot be saved in binary format. You will run into troubles if you try to load this file later. You should clear all YALMIP objects first, or avoid having any YALMIP objects in the list of variables which you
save.
```

Even though  [[1]](https://groups.google.com/g/yalmip/c/maXY14CnUv8) says that YALMIP objects (eg: constraints) cannot be saved, I am just wondering whether there is any way in which I can export `H` from the HPC and use it for plotting purposes.

Thanks in advance.

**Comment by johanlofberg** [comment, marked as answer]
You will have to export the numerical basis and use that for your external purposes

https://yalmip.github.io/command/getbase/

**Reply by gayanlanke** [reply]
Dear Prof. Johan,

Thank you for your prompt reply.

I have a rough idea on how [1](https://yalmip.github.io/command/getvariables/) and [2](https://yalmip.github.io/command/getbase/) works, however, I am not sure how to apply the same concept when it comes to inequality constraints `F1,F2,...` in my source code?

Could you please elaborate this a bit?

Another thing is, the convex hull comprise of approximately 100 inequality constraints corresponding to half spaces. Hence, I am not sure whether `getbase` method works or not?

Thanks.

**Comment by johanlofberg** [comment]
Well, a first problem is that you appear to think that the result from hull is a polytope in (x,y). It is not, it is a polytope (defined by both inequalities and equalities) in lifted (x,y,t)-space, and thus requires a possibly expensive projection to get a polytope in (x,y)

You are much better of constructing the vertices of the polytopes manually, and then construct the hull of that.

or simply catch them from the plot command (where they are generated using approximate ray-casting) if it is only for simple experiments

```
>> V = plot(H,[x;y]);V = V{1};
>> plot(H,[x;y]);hold on
>> plot(V(1,:),V(2,:),'*');
```

**Reply by gayanlanke** [reply]
Dear Prof.

Thank you for your reply.

With the help of your code snippet, I used a small trick to obtain `H` directly without saving or exporting YALMIP variables (MPT3 Toolbox is also needed here).

```
clc;clear

sdpvar x y

F1 = [-1 <= x <= 1, -1 <= y <= 1];  
F2 = [-3 <= x <= -2, -2 <= y <= 2];
F3 = [-5 <= x <= -4, -3 <= y <= 3];
F4 = [-9 <= x <= -6, -5 <= y <= 5];

H=hull(F1,F2,F3,F4);  %% convex hull

%% plot function -->  used to determine vertices of the convex hull --> usually redundant

V = plot(H,[x,y]);
V=V{1};
V_pt = [V(1,:)', V(2,:)']; %% arranging in (x,y) pairs

P=Polyhedron(V_pt); %% convex hull as a convex polygon --> convex hull of convex polytopes

%% checking for redundancies
P.irredundantVRep

%% if redudant, obtain the minimal representation
P_min = P.minVRep

%% calculate the vertices and rays 
P_min.V
P_min.R

```

Once you know the vertices and rays (or else `A` and `b` in the H-representation), the convex polygon can easily be obtained.

I hope this method is correct.

Thanks.

**Reply by johanlofberg** [reply]
If you have MPT you can just as well use MPT for everything (P = Polyhedron(H), and maybe there is a command for obtaining the convex hull from several polyhedral objects Polyhedron(H1))
